### ISSUE TITLE
upload basal rate to Tidepool

### ISSUE THREAD

**Issue by tolot27** [NightscoutFoundation/xDrip#1477]
Currently, basal rate information is not uploaded to Tidepool but is often requested and discussed in [#581 (comment)](https://github.com/NightscoutFoundation/xDrip/issues/581#issuecomment-471898214).

**Comment by pazaan** [comment]
The [code for the data model upload to Tidepool is there](https://github.com/NightscoutFoundation/xDrip/blob/master/app/src/main/java/com/eveningoutpost/dexdrip/tidepool/EBasal.java), so I'm not sure why it's not being uploaded.

**Comment by esayles** [comment]
Thanks for making a separate issue for this. Is anyone aware of any progress or remaining obstacles for this feature?

**Comment by simonpickering** [comment]
My data appear to be uploaded to the standard site (I didn't tick the box to take me to the int- site)

Just as a random note, for those on basal-bolus (or at the very least for me), basal injections entered in XDrip+ appear as standard bolus injections in TidePool, which is a little confusing.

In my view of the data on the website there are only two types of insulin: named "basal" and "bolus", I'm not sure whether the API can work with different types of insulin in a given category. I will do some looking through the docs though.

**Comment by fjpezuela** [comment]
Hello,

A few days ago I connected xDrip with Tidepool. Everything syncs correctly except the basal ones. I have found this thread and the referenced one and it seems that there are users who have worked for it.

Do you know what the problem is and at what point is a possible solution?

Thank you very much for all the support.

**Comment by simonpickering** [comment]
The issue is afaiu on the Tidepool side. I raised the same question on the Tidepool.org #general Slack channel on April 18th at 11:18am (UK time) - they have added in the internal plumbing but it's not yet supported in Blip, which does the uploading afaicr.

With that said, that was a little while ago, so it may now be working in the uploader and requires a mod to the XDrip+ code, I'll need to have a dig unless someone else knows offhand?

If you can't find the Slack thread give me a shout and I'll copy and paste the response in here.

**Comment by fjpezuela** [comment]
Thank you very much for the reply.

I don't have access to Slacks. I would appreciate if you put the answer here.

If I have not misunderstood, making some modifications in xDrip, the basal rise should work.

Hopefully this can be corrected soon.

Thanks again.

**Comment by simonpickering** [comment]
No worries, here's the thread copied and pasted from Slack. As far as I remember a modification will be required in the XDrip+ upload code but only once Tidepool supports it in Blip. I've not check on the status of either - I had a yearly hospital appointment and was trying to work out how to share my data with them. I will have a further dig at some point in the not to distant future to see what the status is.

-------------------------------

Simon Pickering Apr 18th at 11:18 AM
Morning all, I've been looking through the docs to see whether different types of bolus insulin can be recorded. I can't see anything obvious. I'm on basal-bolus with a pair of pens. When uploading from XDrip+ my basal data appears as a bolus and looks identical to the usual bolus injections. I guess the answer here is to add annotations to the injection events.

Simon Pickering  3 months ago
So I've got 3 questions you might be able to help me with: 1) Have I missed a better way of doing this? 2) Should basal insulin which is only injected once a day be uploaded as a basal type (though these all assume a duration over which a quantity will be injected, which would mean fudging the data to make it fit the model) or as a bolus (but it would be nice to mark this as basal and have it appear as such in the web interface - I'm guessing the annotation isn't enough to do this), 3) Can I arbitrarily create annotations or do they have to be listed in annotationdefinitions.js beforehand? Thanks :slightly_smiling_face:

gerrit  3 months ago
Hi @Simon Pickering!
We have created an insulin data type to record injection events, but it's not yet in the public documentation.
I'm not sure if/how we're visualising the insulin data type in Tidepool Web, so I'm checking on that for you.
If you create a new annotation, it will show up with a default text (We cant be 100% certain of the data displayed here.. ) until the definition is added to annotationdefinitions.js.

Simon Pickering  3 months ago
Thanks @gerrit, I've had a dig through the code (blip and the uploader) and can't see any references to an insulin type (or field for the bolus type). Is it in the externally facing repo (and I've just missed it! :))? Thanks

gerrit  3 months ago
Uploader doesn't use the insulin  data type yet. You can see the implementation on the backend here: https://github.com/tidepool-org/platform/tree/master/data/types/insulin
TidepoolKit (the part of Tidepool Loop that sends data to Tidepool) implements it here:
https://github.com/tidepool-org/TidepoolKit/blob/dev/TidepoolKit/Model/Datum/TInsulinDatum.swift
I'm not sure if it's it Blip yet, but I'm trying to find out for you.

Simon Pickering  3 months ago
Great thanks @gerrit :slightly_smiling_face:

gerrit  3 months ago
I got confirmation that we don't yet visualise the insulin data type in Blip, so I guess we're not at a stage yet where you can distinguish between injection events and boluses.

**Comment by fjpezuela** [comment]
Thanks again. See if someone can check it out and readjust the code. I will continue to monitor the progress.

**Comment by akhilshah128** [comment]
Sorry for duplicating this issue

Has there been any update regarding this?

Many thanks

**Comment by janvitos** [comment]
Hi! I was wondering if there's any updates on this ?

We also normally use Tidepool to share our diabetic child's data with his endocrinologist, and having the basal uploaded from xDrip to Tidepool is a must. Since we switched from regular Omnipod to AndroidAPS + xDrip, we no longer have the basal in Tidepool, so we'll have to send his endo the reports generated from Nightscout for now. Hopefully, this will do for the time being.

Thanks!

**Comment by jgslade** [comment]
I just logged into my tidepool and I have basal rates in now. I am using version 2122051214.

**Comment by janvitos** [comment]
> I just logged into my tidepool and I have basal rates in now

Gonna check that out right now! Thanks for the info :)

**Comment by jgslade** [comment]
> > I just logged into my tidepool and I have basal rates in now
> 
> Gonna check that out right now! Thanks for the info :)
> 
> Edit: Unfortunately, even after updating to 2122051214, I don't see the basal rate in Tidepool.
> 
> Anything specific you did in xDrip for the basal rate to show up?

No, I haven't adjusted my settings for a very long time.

**Comment by tolot27** [comment]
This was implemented/activated with 5058163 and released with [2022.04.30](https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.04.30).
@psonnera Please can you add it to the [Changelog](https://xdrip.readthedocs.io/en/latest/changelog)?

Will close this issue since it is resolved.

**Comment by psonnera** [comment]
There are still glitches on the TBR display but Tidepool upload is functional. Documentation in progress.

**Comment by janvitos** [comment]
This is great news! I'm sure tons of people will love this.

On another note, I'm not getting the basal rate neither in xDrip or Tidepool.

Is this supposed to work when combined with AndroidAPS? Because that's where my basal rate is coming from.

Thanks!

**Comment by jgslade** [comment]
> This is great news! I'm sure tons of people will love this.
> 
> On another note, I'm not getting the basal rate neither in xDrip or Tidepool.
> 
> Is this supposed to work when combined with AndroidAPS? Because that's where my basal rate is coming from.
> 
> Anything special I must configure to get the basal rate from AndroidAPS to xDrip?
> 
> Thanks!
I have both options selected in the xdrip+ status(watch) in aaps preferences. I also have all the display stuff in xdrip about basal enabled.

**Comment by janvitos** [comment]
After enabling the first status watch option (IOB), I started seeing a blue dashed bar in xDrip, which is progress I assume. But I'm still not seeing the basal in Tidepool even after a few readings have been uploaded.

I will try enabling the second option in status watch and see how it goes.

**Comment by arpyman** [comment]
@janvitos Please, can tou share of the screen as you set it. Because I can't find it (BGI and IOB settings)

**Comment by janvitos** [comment]
> Please, can you share of the screen as you set it

Sure, here you are. These are the two options that made it work for me:

https://imgur.com/a/A1bJPJI

Cheers.

**Comment by psonnera** [comment]
Just saw the glitches do upload wrongly in Tidepool.

**Comment by tolot27** [comment]
> Just saw the glitches do upload wrongly in Tidepool.

The corresponding issue is #2130. Please continue there.
